src/main/java/org/broadinstitute/hellbender/engine/ReadShard.java

```diff
@@ -2,241 +2,106 @@
 import org.broadinstitute.hellbender.utils.read.GATKRead;
 
-import java.util.*;
-import java.util.stream.Collectors;
-import java.util.stream.StreamSupport;
-
-/**
- * A class to represent a shard of reads data, optionally expanded by a configurable amount of padded data.
- *
- * The reads are lazily loaded by default (when accessing the reads via {@link #iterator}. Loading all the
- * reads in the shard at once is possible via {@link #loadAllReads}.
- *
- * The reads returned will overlap the expanded padded interval. It's possible to query whether they are within
- * the main part of the shard via {@link #contains} and {@link #containsStartPosition}.
- *
- * The reads in the shard can be filtered via {@link #setReadFilter} (no filtering is performed by default).
- */
-public final class ReadShard implements Iterable<GATKRead>, Locatable {
-
-    private final SimpleInterval interval;
-    private final SimpleInterval paddedInterval;
-    private final ReadsDataSource readsSource;
-    private ReadFilter readFilter;
-    private ReadsDownsampler downsampler;
-
-    /**
-     * Create a new ReadShard spanning the specified interval, with the specified amount of padding.
-     *
-     * @param interval the genomic span covered by this shard
-     * @param paddedInterval the span covered by this shard, plus any additional padding on each side (must contain the un-padded interval)
-     * @param readsSource source of reads from which to populate this shard
-     */
-    public ReadShard( final SimpleInterval interval, final SimpleInterval paddedInterval, final ReadsDataSource readsSource ) {
-        Utils.nonNull(interval);
-        Utils.nonNull(paddedInterval);
-        Utils.nonNull(readsSource);
-        Utils.validateArg(paddedInterval.contains(interval), "The padded interval must contain the un-padded interval");
-
-        this.interval = interval;
-        this.paddedInterval = paddedInterval;
-        this.readsSource = readsSource;
-    }
-
-    /**
-     * Create a new ReadShard spanning the specified interval, with no additional padding
-     *
-     * @param interval the genomic span covered by this shard
-     * @param readsSource source of reads from which to populate this shard
-     */
-    public ReadShard( final SimpleInterval interval, final ReadsDataSource readsSource ) {
-        this(interval, interval, readsSource);
-    }
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
 
-    /**
-     * Reads in this shard will be filtered using this filter before being returned.
-     * Read filtering will be performed before any requested downsampling.
-     *
-     * @param filter filter to use (may be null, which signifies that no filtering is to be performed)
-     */
-    public void setReadFilter( final ReadFilter filter ) {
-        this.readFilter = filter;
-    }
+public interface ReadShard extends Iterable<GATKRead>,Locatable {
```

Comment:
You can also have a look to https://github.com/broadinstitute/gatk/pull/1708 for the interface of shard data, because I think that it will make more sense to have an interface like the `SourceShard` as I implemented it there. I can extract that classes for my branch if they are easier to review and to add them to the GATK framework before this PR is accepted...